// @1password/sdk/dist/secrets.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SecretsSource_inner;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsSource = void 0;
class SecretsSource {
    constructor(inner) {
        _SecretsSource_inner.set(this, void 0);
        __classPrivateFieldSet(this, _SecretsSource_inner, inner, "f");
    }
    /**
     *  Resolve returns the secret the provided secret reference points to.
     */
    resolve(secretReference) {
        return __awaiter(this, void 0, void 0, function* () {
            const invocationConfig = {
                clientId: __classPrivateFieldGet(this, _SecretsSource_inner, "f").id,
                invocation: {
                    name: "SecretsResolve",
                    parameters: {
                        secret_reference: secretReference,
                    },
                },
            };
            return JSON.parse(yield __classPrivateFieldGet(this, _SecretsSource_inner, "f").core.invoke(invocationConfig));
        });
    }
}
exports.SecretsSource = SecretsSource;
_SecretsSource_inner = new WeakMap();
