// @1password/sdk/dist/types.js
"use strict";
/*
 Generated by typeshare 1.10.0-beta.7
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemFieldType = exports.ItemCategory = void 0;
var ItemCategory;
(function (ItemCategory) {
    ItemCategory["Login"] = "Login";
    ItemCategory["SecureNote"] = "SecureNote";
    ItemCategory["CreditCard"] = "CreditCard";
    ItemCategory["CryptoWallet"] = "CryptoWallet";
    ItemCategory["Identity"] = "Identity";
    ItemCategory["Password"] = "Password";
    ItemCategory["Document"] = "Document";
    ItemCategory["ApiCredentials"] = "ApiCredentials";
    ItemCategory["BankAccount"] = "BankAccount";
    ItemCategory["Database"] = "Database";
    ItemCategory["DriverLicense"] = "DriverLicense";
    ItemCategory["Email"] = "Email";
    ItemCategory["MedicalRecord"] = "MedicalRecord";
    ItemCategory["Membership"] = "Membership";
    ItemCategory["OutdoorLicense"] = "OutdoorLicense";
    ItemCategory["Passport"] = "Passport";
    ItemCategory["Rewards"] = "Rewards";
    ItemCategory["Router"] = "Router";
    ItemCategory["Server"] = "Server";
    ItemCategory["SshKey"] = "SshKey";
    ItemCategory["SocialSecurityNumber"] = "SocialSecurityNumber";
    ItemCategory["SoftwareLicense"] = "SoftwareLicense";
    ItemCategory["Person"] = "Person";
    ItemCategory["Unsupported"] = "Unsupported";
})(ItemCategory || (exports.ItemCategory = ItemCategory = {}));
var ItemFieldType;
(function (ItemFieldType) {
    ItemFieldType["Text"] = "Text";
    ItemFieldType["Concealed"] = "Concealed";
    ItemFieldType["CreditCardType"] = "CreditCardType";
    ItemFieldType["Phone"] = "Phone";
    ItemFieldType["Url"] = "Url";
    ItemFieldType["Totp"] = "Totp";
    ItemFieldType["Unsupported"] = "Unsupported";
})(ItemFieldType || (exports.ItemFieldType = ItemFieldType = {}));
